numpy/ma/core.py

```diff
@@ -3708,7 +3730,13 @@ def filled(self, fill_value=None):
         if fill_value is None:
             fill_value = self.fill_value
         else:
-            fill_value = _check_fill_value(fill_value, self.dtype)
+            if ((np.isscalar(fill_value) == True) or (self.shape == fill_value.shape)):
```

Comment:
Its pretty tricky to do such a check. It probably makes sense to move this check into the `_check_fill_value` function. There are some tricky things about object arrrays (arrays with `dtype=object`). If we do not want to worry about it for now too much (it may be a different issue), what I think will work is: Adding an `fill_value.ndim == 0` check at the end of `_check_fill_value`. You could refactor the last `np.array` call, since it seems unnecessary except in the default fill value branch.

Reply:
If we keep that as a final check, a number of tests fail, which are passing non-scalar values as the fill_value. That's why, even in the above if statement, I let it be true even when the fill_value were non-scalar. as long as their shape was the same as that of the input array. So from here, either we have to modify a lot of tests or develop some other workaround. What do you suggest?

Reply:
oh OK, I missed that shape check. `fill_value` should be an array here though, so checking `.ndim == 0 or shape == fill_value.shape` may be an option? Depending on where else `_check_fill_value` is called, moving it inside and with a new shape argument may or may not make sense.